Team driving support

pnc opened the pull request:
This implements the design from opencab#17 in the sample
consumer and provider.

At a high level, it provides a way for:

1. Identity providers to indicate when a second (or third, etc.) driver
   has logged into the device.
2. HOS providers to return HOS information for each logged-in user,
   including duty status and hours-of-service clocks.

It aims to be backwards compatible, so that single-driver functionality
will continue to work between an older consumer app and a newer provider
or vice-versa.

pnc committed:
Add new durationSeconds property to HOSContract.Clock

pnc committed:
Bump HOSContract version to 0.3 for clock backwards compatibility

Additional testing of opencab#19 revealed that although `Parcel` is capable
of reading a single parceled object from a newer version that adds
additional fields to the end (in this case, `durationSeconds`), since
the `Clock` objects are in an array, the increased length of the
unconsumed field throws off the offset of subsequent objects in the
array and leads to a crash inside the `Parcel` code.

To avoid causing such a crash in a pre-existing HOS consumer that only
understands version 0.2 clocks (no `durationSeconds`) when it encounters
values from a version 0.3 provider, we are bumping the version number so
that:

  1. New consumers MUST explicitly indicate support for `durationClocks` by
     passing `"0.3"` to the `version` parameter of `ContractProvider.call`
     when fetching HOS status.
  2. New providers MUST NOT return a version 0.3 clock (with a
     `durationSeconds` value) if the calling consumer has not passed a
     version equal to or greater than `"0.3"`. New providers MAY return
     an error if they do not need or want to support older consumers.
  3. All new providers and consumers SHOULD support version 0.3 so that
     support for 0.2 can eventually be dropped.

In addition, the fact that `Parcel` can't natively handle new values
in a backwards-compatible way means we'll stop using it for all future
contracts or contract versions, with the exception of the in-flight
opencab#17 work.

We need to define "greater than" for version numbers, presumably using
non-lexical comparison.

Co-authored-by: Victor Ceban <victor.ceban@eleostech.com>

pnc committed:
Remove defunct Parcel backwards-compat test

VCEleosTech committed:
Update Gradle and build tools version

VCEleosTech committed:
Update examples to show how providers need to enumerate receiver packages

The previous sample consumer had its receiver classes all named at the top
level of the app's package. This resulted in one implementation that
inadvertently depended on the receiver class not being in a sub-package,
which broke compatibility.

The existing sample provider had code to enumerate packages, but this
updates the consumer to actually exercise this by moving the receiver
package to a sub-package. It also unifies this logic across all
broadcast types.

VCEleosTech committed:
Add support for team drivers to Identity and HOS contracts

This implements the design from opencab#17 in the sample
consumer and provider.

At a high level, it provides a way for:

1. Identity providers to indicate when a second (or third, etc.) driver
   has logged into the device.
2. HOS providers to return HOS information for each logged-in user,
   including duty status and hours-of-service clocks.

It aims to be backwards compatible, so that single-driver functionality
will continue to work between an older consumer app and a newer provider
or vice-versa.

ZonarTimH commented:
There will be two Zonar reviewers on this.  My team handles the IdentityContract portion.  I asked the devs to double-check my read.  They came back with a thumbs-up.  Zonar is fine with the IdentityContract piece of the change.

I can't speak for the HOSContract piece, which is handled by another team.  I've forwarded this to Quynh Nguyen, who manages that team.  His signoff is also needed before Zonar has fully approved.

LMK if you need anything further from Zonar!

ZonarTimH reviewed (commented):
I said this in MR comments.  Reposting in case this routes to a different place:

At Zonar, IdentityContract and HOSContract are implemented by two different teams.  I manage the folks who implement IdentityContract.  We have reviewed the IdentityContract change, and we approve it.

The HOSContract implementers are managed by Quynh Nguyen.  I forwarded this to him at the beginning of the week.  Hopefully you get word from him soon.

Either way, please keep us all posted on when you want this change to go to the world.  Zonar will need to figure out the timing of a Shell release which implements these changes.  Thanks!

pnc commented:
Thanks a ton @ZonarTimH!

Fixing the docs so they can be easily browsed is still on my list.

pnc committed:
Add version comparison sample code

This adds sample code (Version.java) for comparing version numbers and
modifies the sample provider and consumer to examine the version
information that is sent and received to demonstrate how backwards-
and forwards-compatibility work.

pnc committed:
Fix: Remove durationSeconds from Clock (0.2) class

pnc committed:
Exercise more clock types in example provider data

pnc committed:
Remove bogus view bindings

pnc committed:
Update builds tools

pnc committed:
Version bump samples 1.0 -> 1.1

pnc committed:
Fix javadoc errors

pnc committed:
Set LANG when building

pnc committed:
Install JDK second in CI

pnc commented on OpenCabConsumer/app/src/main/java/com/eleostech/exampleconsumer/MainActivity.java:
```diff
@@ -91,6 +97,78 @@ private void logEvent(String event) {
+                                if (result.containsKey(HOSContract.KEY_VERSION) && result.get(HOSContract.KEY_VERSION) != null) {
+                                    resultVersion = new Version(result.getString(HOSContract.KEY_VERSION));
+                                } else {
+                                    resultVersion = new Version("0.3");
```
I think this should be `0.2` per 4.3.4 but I might be missing something!

pnc commented on OpenCabConsumer/app/src/main/java/com/eleostech/exampleconsumer/MainActivity.java:
```diff
@@ -106,7 +184,7 @@ private void callVehicleInformationProvider() {
+                                result = resolver.call(authority, VehicleInformationContract.METHOD_GET_VEHICLE_INFORMATION, VehicleInformationContract.VERSION, null);
                             } catch (Exception ex) {
                                 Log.i(LOG_TAG, "Error calling provider: ", ex);
                                 adapterVehicleInformation.insert(dateTime + " : " + "Package: " + provider.packageName + ", Error: " + ex.getMessage(), 0);
```
This should implement a version check per 4.3.2.

pnc commented on OpenCabConsumer/app/src/main/java/com/eleostech/exampleconsumer/MainActivity.java:
```diff
@@ -161,18 +239,27 @@ private void callIdentityProviderGetLoginCredentials() {
-                                        adapterLoginCredentials.insert(dateTime + " : " + "Package: " + provider.packageName  + ", Token: " + loginCredentials.getToken() + ", Provider: " + loginCredentials.getProvider(), 0);
-                                    } else {
-                                        adapterLoginCredentials.insert(dateTime + " : " + "Package: " + provider.packageName + ", Error: No login credentials found", 0);
+                                if (result.containsKey(IdentityContract.KEY_LOGIN_CREDENTIALS) || result.containsKey(IdentityContract.KEY_ALL_LOGIN_CREDENTIALS)) {
```
This should implement a version check per 4.3.2.

pnc commented on OpenCabConsumer/app/src/main/java/com/eleostech/exampleconsumer/MainActivity.java:
```diff
@@ -214,7 +301,9 @@ private void callIdentityProviderGetActiveDrivers() {
                                     List<IdentityContract.Driver> drivers = result.getParcelableArrayList(IdentityContract.KEY_ACTIVE_DRIVERS);
                                     Log.d(LOG_TAG, "Got drivers: " + drivers);
                                     if (drivers != null && drivers.size() > 0) {
-                                        adapterActiveDrivers.insert(dateTime + " : " + "Package: " + provider.packageName + ", Driver Name: " + drivers.get(0).getUsername() + ", Is Driving: " + drivers.get(0).isDriving(), 0);
```
This should implement a version check per 4.3.2.

VCEleosTech committed:
removing IDE files as they are not supposed to be checked into repo

VCEleosTech committed:
Update provider implementations to set KEY_VERSION to indicate the version of their response
https://3.basecamp.com/4496203/buckets/32502950/todos/6391766484

pnc committed:
Clarify that getManageAction and getLogoutAction need to be URIs

pnc commented:
@Mehantp @BoKlassen Please see the latest commit, which fixes the docs for the HOS "manage action" and "logout action" to specify that it's a URI, not an "intent string." (Android seems to have pretty completely eradicated the old `intent:` URI scheme for security reasons, so I don't think it's a good idea to pin the spec to that.)

https://github.com/opencabstandard/opencab/pull/21/commits/ae1a215989eab7a2374a9554dcc8d2d5e4d93b8f

pnc committed:
Restore setClocks (mistakenly removed)

BoKlassen commented:
> @Mehantp @BoKlassen Please see the latest commit, which fixes the docs for the HOS "manage action" and "logout action" to specify that it's a URI, not an "intent string." (Android seems to have pretty completely eradicated the old `intent:` URI scheme for security reasons, so I don't think it's a good idea to pin the spec to that.)
> 
> [ae1a215](https://github.com/opencabstandard/opencab/commit/ae1a215989eab7a2374a9554dcc8d2d5e4d93b8f)

Awesome, thanks. This works better for us.

pnc commented:
Good deal, that matches real-world usage of that property today, so it should be net no change for @ZonarTimH et al.

It does leave open the possibility of users seeing a disambiguation dialog if two apps try to use the register for the same URI scheme in their manifests, but I think that risk is fairly remote.

pnc committed:
Fix @link javadoc refs

pnc committed:
Add example sequence for team driver SSO

pnc committed:
Add sample interactions for HOS contract 0.3

pnc committed:
Update docs examples for HOS 0.3

pnc committed:
Parelable -> Parcelable typo

pnc committed:
Clarify backwards/forwards compatibility behavior in spec

VCEleosTech committed:
Sample provider must never return new key (team HOS) for 0.2 (#35)

https://3.basecamp.com/4496203/buckets/32502950/todos/6407097832#__recording_6407244515

pnc committed:
Fix some HTML tags

VCEleosTech committed:
KEY_TEAM_HOS returns an Array of HOSStatusV2 (#36)

https://3.basecamp.com/4496203/buckets/32502950/messages/6374919029#__recording_6407351887

pnc commented on OpenCabProvider/app/src/main/java/org/opencabstandard/provider/package-info.java:
```diff
@@ -145,9 +146,38 @@
+ * since last launch will be started by the OS and given an opportunity to handle the event.</li>
+ * </ol>
+ *
+ * <h4>3.5. Backwards- and forwards-compatibility</h4>
```
@ZonarTimH @BoKlassen @Mehantp We've added a bunch of info here about version interop. Our intention is to clarify but not make big changes/cause rework for your teams.

Does everything here track with your vision and/or implementations? (In particular, we had to update our code and the provider samples to better handle requests from newer consumers.)

ZonarTimH replied:
I'll get that double-checked w the team. Please note that there are two approvers at Zonar, myself for IdentityContract and Quynh Nguyen for HOSContract.  I think it'd be good for you to get Quynh in here directly, vs me alerting him when there's something to review.

pnc replied:
@ZonarTimH Can do. Do you have Quynh's GitHub username? I can add to the reviewers list!

ZonarTimH replied:
Sure, just checked.  He is @quynh_nguyen.  Thanks!

ZonarTimH replied:
Pardon me, Quỳnh prefers his khongtu username.  I think I gave you gitlab above.  Go with khongtu.

ZonarTimH replied:
Deleted my earlier comment so there's no confusion about usernames.

pnc replied:
Thanks! Due to the magic of email notifications, I missed the deletion but have now invited both usernames to the repository's review/working group.

ZonarTimH replied:
I got that double-check with the team.  Sorry on the holdup.  The result is a green light.  Zonar is OK with the revised IdentityContract changes.

pnc committed:
Remove CSS noise

pnc committed:
Add some references to the spec in the sample implementations

pnc reviewed (commented):
Okay, sorry for the commit noise but I think this should be ready for merge.

@BoKlassen and @ZonarTimH, in the absence of a formal governance board meeting or something, would you both be willing to approve if this all checks out from your side?

scott-parker4 committed:
Add toggle to change logout action

pnc committed:
Merge pull request #23 from opencabstandard/logout-action-toggle

Add toggle to change logout action

pnc commented:
This resolves #17.
